src/ATen/native/xpu/sycl/SortingRadixSelect.h

```diff
@@ -0,0 +1,454 @@
+
+  // Converts a float to an integer representation with the same
+  // sorting; i.e., for floats f1, f2:
+  // if f1 < f2 then convert(f1) < convert(f2)
+  // We use this to enable radix selection of floating-point values.
+  // This also gives a relative order for NaNs, but that's ok, as they
+  // will all be adjacent
+  // neg inf: signbit=1 exp=ff fraction=0 --> radix = 0 00 ff..
+  // pos inf: signbit=0 exp=ff fraction=0 --> radix = 1 ff 00..
+  // pos nan: signbit=0 exp=ff fraction>0 --> radix = 1 ff x>0
+  // neg nan: signbit=1 exp=ff fraction>0 --> radix = 0 00 x<ff...
+  static inline RadixType convert(float v) {
+    RadixType x = *((uint32_t*)&v);
+    RadixType mask = (x & 0x80000000) ? 0xffffffff : 0x80000000;
+    return (x ^ mask);
+  }
+
+  static inline float deconvert(RadixType v) {
+    RadixType mask = (v & 0x80000000) ? 0x80000000 : 0xffffffff;
+    return __int_as_float(v ^ mask);
+  }
+};
+
+template <>
+struct TopKTypeConfig<uint8_t> {
+  using RadixType = uint32_t;
+
+  static inline RadixType convert(uint8_t v) {
+    return v;
+  }
+
+  static inline uint8_t deconvert(RadixType v) {
+    return v;
+  }
+};
+
+template <>
+struct TopKTypeConfig<int8_t> {
+  using RadixType = uint32_t;
+
+  static inline RadixType convert(int8_t v) {
+    return 128u + v;
+  }
+
+  static inline int8_t deconvert(RadixType v) {
+    return v - 128;
+  }
+};
+
+template <>
+struct TopKTypeConfig<int16_t> {
+  using RadixType = uint32_t;
+
+  static inline RadixType convert(int16_t v) {
+    return 32768u + v;
+  }
+
+  static inline int16_t deconvert(RadixType v) {
+    return v - 32768;
+  }
+};
+
+template <>
+struct TopKTypeConfig<int32_t> {
+  using RadixType = uint32_t;
+
+  static inline RadixType convert(int32_t v) {
+    return 2147483648u + v;
+  }
+
+  static inline int32_t deconvert(RadixType v) {
+    return v - 2147483648u;
+  }
+};
+
+template <>
+struct TopKTypeConfig<int64_t> {
+  using RadixType = uint64_t;
+
+  static inline RadixType convert(int64_t v) {
+    return 9223372036854775808ull + v;
+  }
+
+  static inline int64_t deconvert(RadixType v) {
+    return v - 9223372036854775808ull;
+  }
+};
+
+template <>
+struct TopKTypeConfig<double> {
+  using RadixType = uint64_t;
+
+  static inline RadixType convert(double v) {
+    RadixType x = *((uint64_t*)&v);
+    RadixType mask = -((x >> 63)) | 0x8000000000000000;
+    return (x ^ mask);
+  }
+
+  static inline double deconvert(RadixType v) {
+    RadixType mask = ((v >> 63) - 1) | 0x8000000000000000;
+    return __long_long_as_double(v ^ mask);
+  }
+};
+
+template <>
+struct TopKTypeConfig<at::Half> {
+  using RadixType = uint32_t;
+
+  static inline RadixType convert(at::Half v) {
+    RadixType x = *((uint16_t*)&v);
+    RadixType mask = -((x >> 15)) | 0x8000;
+    return (x ^ mask);
+  }
+
+  static inline at::Half deconvert(RadixType v) {
+    RadixType mask = ((v >> 15) - 1) | 0x8000;
+    auto v_de = v ^ mask;
+    return *((at::Half*)&v_de);
```

Comment:
got warning here "warning: ‘v_de’ is used uninitialized [-Wuninitialized]"
I am considering replace with union